### PR TITLE
Add EPVector bounds checking for debug builds

### PR DESCRIPTION
**Pull request by jasondegraw**
Pull request overview
---------------------
There was discussion of bounds checking of vectors and arrays on (if I recall correctly) the last iteration call. This is what it would look like if we added a bounds checks to `EPVector` for debug builds. On debug builds, both `operator[]` and `operator()` get routed through `at` of the base `std::vector`, which throws for out-of-range indices. This won't change what happens with `std::array`. If we wanted something similar there, we could define an `EPArray` that would be an alias for `std::array` on release builds and would do something similar to `EPVector` for debug builds. I don't know if that's really a good idea. With `std::array` being relatively new to the codebase, it might be better to work on ways to use those better and avoid situations where indices can get out of range.

This PR also adds an `EPVector` unit test.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/EPVector.hh [hunk 1]
```diff
@@ -58,7 +58,9 @@ namespace EnergyPlus {
 template <typename T> struct EPVector : private std::vector<T>
 {
     using std::vector<T>::size;
+#ifdef NDEBUG
     using std::vector<T>::operator[];
```
Comment (jasondegraw) [inline src/EnergyPlus/EPVector.hh:62]: On release builds, use the base `operator[]` and act just like `std::vector`.

File: src/EnergyPlus/EPVector.hh [hunk 2]
```diff
@@ -72,6 +74,18 @@ template <typename T> struct EPVector : private std::vector<T>
     using value_type = T;
     using size_type = typename std::vector<T>::size_type;
 
+#ifndef NDEBUG
+    [[nodiscard]] T &operator[](std::size_t n)
+    {
+        return std::vector<T>::at(n);
```
Comment (jasondegraw) [inline src/EnergyPlus/EPVector.hh:80]: For debug builds, use the base `at` member function, which throws for bad indices.

File: src/EnergyPlus/EPVector.hh [hunk 3]
```diff
@@ -126,7 +140,9 @@ private:
 template <> struct EPVector<bool> : private std::vector<std::uint8_t>
```
Comment (jasondegraw) [inline src/EnergyPlus/EPVector.hh:140]: Boolean vectors require a little more work, which is not surprising, but it's basically the same as the above with a few additional changes to accommodate the `noexcept` labeling.

**Comment by Myoldmopar** [comment]
What's the status of this one?  Hoping to freeze today so I'm happy to pull in develop and do any testing to get it in if that's the expectation.

**Comment by jasondegraw** [comment]
I'd prefer to have a review from @lefticus or @mbadams5 of the `bool` special case, but as far as I know this is ready to go.

**Comment by Myoldmopar** [comment]
I don't know that @lefticus will have time to review it today, perhaps @mbadams5 could.  Also if we could extend the unit test to cover the bool special case, that would make be feel better about it.  Do you want me to do that?

**Comment by mbadams5** [comment]
I am looking over this but it'll likely be later this evening before I can finish my review

**Comment by Myoldmopar** [comment]
Thanks @mbadams5, I did make a copy of the EPVector unit test and replaced it with a vector of bool and it works with no problem.  I'm happy to bulk it up if there is some specific behavior we'd be looking for to ensure the implementation is robust enough.

**Review by mbadams5 — APPROVED** [review]
This looks good to me. I'd say pull in develop and push up your unit test changes @Myoldmopar.

**Comment by Myoldmopar** [comment]
Done, thanks @mbadams5

**Comment by Myoldmopar** [comment]
All good here, merging this.  Thanks @jasondegraw

**Pull request merged**